File: lessons/lesson_02_memory_structs/cpp/src/main.cpp
#include <Arduino.h>

// 1. Структура даних сенсора
struct SensorData {
  float temperature;   // Температура у °C (15.0 - 30.0)
  float humidity;      // Вологість у % (30.0 - 65.0)
  uint32_t timestamp;  // Час роботи системи в секундах (uptime)
};

// Періодичні інтервали для задач (в мілісекундах)
const unsigned long SENSOR_READ_INTERVAL = 20000;  // 20 секунд
const unsigned long HEAP_MONITOR_INTERVAL = 60000; // 60 секунд (1 хвилина)

// Змінні таймерів для неблокуючого виконання (non-blocking timing)
unsigned long lastSensorReadTime = 0;
unsigned long lastHeapMonitorTime = 0;

// Змінна для відстеження початкового значення вільної пам'яті
uint32_t initialFreeHeap = 0;

// Функція для генерації випадкових даних сенсора
SensorData readSensorData() {
  SensorData data;
  // Генерація температури в діапазоні 15.0 ... 30.0 °C
  data.temperature = random(150, 301) / 10.0f;
  // Генерація вологості в діапазоні 30.0 ... 65.0 %
  data.humidity = random(300, 651) / 10.0f;
  // Timestamp як поточний час роботи програми у секундах від старту
  data.timestamp = millis() / 1000;
  return data;
}

// Функція для демонстрації переповнення типів даних (Overflow)
void demonstrateIntegerOverflow() {
  Serial.println("==================================================");
  Serial.println("1. Демонстрація переповнення типів даних (Integer Overflow)");
  Serial.println("==================================================");

  uint8_t a = 200;
  uint8_t b = 100;
  uint8_t sum = a + b;

  Serial.print("Змінна a (uint8_t): "); Serial.println(a);
  Serial.print("Змінна b (uint8_t): "); Serial.println(b);
  Serial.print("Математична сума (a + b): "); Serial.println((int)a + (int)b);
  Serial.print("Результат у sum (uint8_t): "); Serial.println(sum);
  Serial.println("--------------------------------------------------");
  Serial.println("ПОЯСНЕННЯ:");
  Serial.println("Тип uint8_t приймає значення від 0 до 255 (2^8 = 256 станів).");
  Serial.println("При додаванні 200 + 100 виникає overflow: 300 % 256 = 44.");
  Serial.println("==================================================\n");
}

// Вивід даних сенсора
void logSensorData() {
  SensorData currentData = readSensorData();
  Serial.print("[SENSOR DATA] Timestamp: ");
  Serial.print(currentData.timestamp);
  Serial.print("s | Температура: ");
  Serial.print(currentData.temperature, 1);
  Serial.print(" °C | Вологість: ");
  Serial.print(currentData.humidity, 1);
  Serial.println(" %");
}

// Моніторинг вільної пам'яті (Heap)
void logHeapMemory() {
  uint32_t currentFreeHeap = ESP.getFreeHeap();
  int32_t heapDiff = (int32_t)currentFreeHeap - (int32_t)initialFreeHeap;

  Serial.println("--------------------------------------------------");
  Serial.print("[MEMORY MONITOR] Free Heap: ");
  Serial.print(currentFreeHeap);
  Serial.print(" bytes | Min Free Heap: ");
  Serial.print(ESP.getMinFreeHeap());
  Serial.print(" bytes | Delta: ");
  if (heapDiff >= 0) Serial.print("+");
  Serial.print(heapDiff);
  Serial.println(" bytes");

  if (heapDiff < 0) {
    Serial.println("⚠️ УВАГА: Вільна пам'ять зменшується! Можливий витік пам'яті (Memory Leak).");
  } else {
    Serial.println("✅ Стан пам'яті стабільний (Memory consumption is stable).");
  }
  Serial.println("--------------------------------------------------");
}

void setup() {
  Serial.begin(115200);
  delay(1000); // Час на ініціалізацію послідовного порту

  // Ініціалізація генератора випадкових чисел
  randomSeed(analogRead(34));

  Serial.println("\n=== ESP32 Lesson 2: Memory, Structs & Data Types ===");
  
  // 1. Пояснення переповнення типів
  demonstrateIntegerOverflow();

  initialFreeHeap = ESP.getFreeHeap();
  Serial.print("Початковий обсяг вільного Heap: ");
  Serial.print(initialFreeHeap);
  Serial.println(" bytes\n");

  Serial.println("Запуск неблокуючого моніторингу...");
  Serial.println("- Читання сенсора кожні 20 секунд");
  Serial.println("- Перевірка вільної пам'яті кожні 60 секунд\n");

  // Виведемо перше значення відразу при старті
  logSensorData();
  lastSensorReadTime = millis();
  lastHeapMonitorTime = millis();
}

void loop() {
  unsigned long currentMillis = millis();

  // Задача 1: Вивід даних сенсора кожні 20 секунд (20 000 мс)
  if (currentMillis - lastSensorReadTime >= SENSOR_READ_INTERVAL) {
    lastSensorReadTime = currentMillis;
    logSensorData();
  }

  // Задача 2: Моніторинг вільної пам'яті кожні 60 секунд (60 000 мс)
  if (currentMillis - lastHeapMonitorTime >= HEAP_MONITOR_INTERVAL) {
    lastHeapMonitorTime = currentMillis;
    logHeapMemory();
  }
}

File: lessons/lesson_02_memory_structs/rust/src/main.rs
#![no_std]
#![no_main]

use esp_backtrace as _;
use esp_hal::{delay::Delay, main, time::Instant};
use esp_println::println;

// 1. Структура даних сенсора в Rust
#[derive(Debug)]
struct SensorData {
    temperature: f32, // 15.0..30.0 °C
    humidity: f32,    // 30.0..65.0 %
    timestamp: u32,   // Uptime у секундах
}

// Простий генератор псевдовипадкових чисел (Linear Congruential Generator) для no_std
struct SimpleLcg {
    state: u32,
}

impl SimpleLcg {
    fn new(seed: u32) -> Self {
        Self { state: seed }
    }

    fn next_u32(&mut self) -> u32 {
        self.state = self.state.wrapping_mul(1664525).wrapping_add(1013904223);
        self.state
    }

    fn range_f32(&mut self, min: f32, max: f32) -> f32 {
        let step = (self.next_u32() % 1000) as f32 / 1000.0;
        min + step * (max - min)
    }
}

// Демонстрація переповнення типів у Rust
fn demonstrate_integer_overflow() {
    println!("\r==================================================");
    println!("\r1. Демонстрація переповнення типів даних у Rust");
    println!("\r==================================================");

    let a: u8 = 200;
    let b: u8 = 100;
    
    // В Rust переповнення за замовчуванням під час звичайної математики a + b у Debug режимі
    // викликає panic для гарантії безпеки пам'яті. Для явного переповнення використовують wrapping_add.
    let sum_wrapped = a.wrapping_add(b);
    let sum_promoted = (a as u16) + (b as u16);

    println!("\rЗмінна a (u8): {}", a);
    println!("\rЗмінна b (u8): {}", b);
    println!("\rМатематична сума (u16 promotion): {}", sum_promoted);
    println!("\rЯвне переповнення (a.wrapping_add(b)): {}", sum_wrapped);
    println!("\r--------------------------------------------------");
    println!("\rПОЯСНЕННЯ (Rust):");
    println!("\rТип u8 вміщує значення 0..255.");
    println!("\rУ Rust 'a.wrapping_add(b)' виконує циклічне переповнення: 300 % 256 = 44.");
    println!("\rУ Debug-збірках звичайний '+' перевіряє переповнення та захищає від помилок.");
    println!("\r==================================================");
}

#[main]
fn main() -> ! {
    let _peripherals = esp_hal::init(esp_hal::Config::default());
    let delay = Delay::new();

    println!("\r\n=== ESP32 Lesson 2: Memory & Structs (Rust no_std) ===");

    // 1. Пояснення переповнення типів
    demonstrate_integer_overflow();

    let mut rng = SimpleLcg::new(0xDEADBEEF);
    let start_time = Instant::now();

    let mut last_sensor_secs: u64 = 0;
    let mut last_heap_secs: u64 = 0;

    println!("\r");
    println!("\r[RUST MEMORY INFO]");
    println!("\rno_std Rust використовує статичне виділення пам'яті (без Heap allocator).");
    println!("\rВитоки пам'яті (Memory Leaks) гарантовано відсутні на рівні компілятора!");
    println!("\r");

    loop {
        let elapsed_secs = start_time.elapsed().as_secs();

        // Читання та вивід сенсорів кожні 20 секунд
        if elapsed_secs >= last_sensor_secs + 20 || last_sensor_secs == 0 {
            last_sensor_secs = if elapsed_secs == 0 { 1 } else { elapsed_secs };

            let data = SensorData {
                temperature: rng.range_f32(15.0, 30.0),
                humidity: rng.range_f32(30.0, 65.0),
                timestamp: elapsed_secs as u32,
            };

            println!(
                "\r[SENSOR DATA] Timestamp: {}s | Температура: {:.1} °C | Вологість: {:.1} %",
                data.timestamp, data.temperature, data.humidity
            );
        }

        // Перевірка пам'яті кожні 60 секунд
        if elapsed_secs >= last_heap_secs + 60 || (last_heap_secs == 0 && elapsed_secs > 0) {
            last_heap_secs = elapsed_secs;

            println!("\r--------------------------------------------------");
            println!(
                "\r[MEMORY MONITOR] Uptime: {}s | Dynamic Heap Allocations: 0 bytes (no_std)",
                elapsed_secs
            );
            println!("\r✅ Стан пам'яті повністю стабільний (Static Safety Guarantee).");
            println!("\r--------------------------------------------------");
        }

        // Невеликий відпочинок циклу (100 мс)
        delay.delay_millis(100);
    }
}
